File: Firebase/Database/Core/FSparseSnapshotTree.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FSparseSnapshotTree.h"
#import "FChildrenNode.h"

@interface FSparseSnapshotTree () {
    id<FNode> value;
    NSMutableDictionary* children;
}

@end

@implementation FSparseSnapshotTree

- (id) init {
    self = [super init];
    if (self) {
        value = nil;
        children = nil;
    }
    return self;
}

- (id<FNode>) findPath:(FPath *)path {
    if (value != nil) {
        return [value getChild:path];
    } else if (![path isEmpty] && children != nil) {
        NSString* childKey = [path getFront];
        path = [path popFront];
        FSparseSnapshotTree* childTree = children[childKey];
        if (childTree != nil) {
            return [childTree findPath:path];
        } else {
            return nil;
        }
    } else {
        return nil;
    }
}

- (void) rememberData:(id<FNode>)data onPath:(FPath *)path {
    if ([path isEmpty]) {
        value = data;
        children = nil;
    } else if (value != nil) {
        value = [value updateChild:path withNewChild:data];
    } else {
        if (children == nil) {
            children = [[NSMutableDictionary alloc] init];
        }

        NSString* childKey = [path getFront];
        if (children[childKey] == nil) {
            children[childKey] = [[FSparseSnapshotTree alloc] init];
        }

        FSparseSnapshotTree* child = children[childKey];
        path = [path popFront];
        [child rememberData:data onPath:path];
    }
}

- (BOOL) forgetPath:(FPath *)path {
    if ([path isEmpty]) {
        value = nil;
        children = nil;
        return YES;
    } else {
        if (value != nil) {
            if ([value isLeafNode]) {
                // non-empty path at leaf. the path leads to nowhere
                return NO;
            } else {
                id<FNode> tmp = value;
                value = nil;

                [tmp enumerateChildrenUsingBlock:^(NSString *key, id<FNode> node, BOOL *stop) {
                    [self rememberData:node onPath:[[FPath alloc] initWith:key]];
                }];

                // we've cleared out the value and set children. Call ourself again to hit the next case
                return [self forgetPath:path];
            }
        } else if (children != nil) {
            NSString* childKey = [path getFront];
            path = [path popFront];

            if (children[childKey] != nil) {
                FSparseSnapshotTree* child = children[childKey];
                BOOL safeToRemove = [child forgetPath:path];
                if (safeToRemove) {
                    [children removeObjectForKey:childKey];
                }
            }

            if ([children count] == 0) {
                children = nil;
                return YES;
            } else {
                return NO;
            }
        } else {
            return YES;
        }
    }
}

- (void) forEachTreeAtPath:(FPath *)prefixPath do:(fbt_void_path_node)func {
    if (value != nil) {
        func(prefixPath, value);
    } else {
        [self forEachChild:^(NSString* key, FSparseSnapshotTree* tree) {
            FPath* path = [prefixPath childFromString:key];
            [tree forEachTreeAtPath:path do:func];
        }];
    }
}


- (void) forEachChild:(fbt_void_nsstring_sstree)func {
    if (children != nil) {
        for (NSString* key in children) {
            FSparseSnapshotTree* tree = [children objectForKey:key];
            func(key, tree);
        }
    }
}


@end

File: FirebaseDatabaseSwiftCore/Sources/Core/FSparseSnapshotTree.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 07/04/2022.
//

import Foundation

@objc public class FSparseSnapshotTree: NSObject {
    var value: FNode?
    var children: [String: FSparseSnapshotTree] = [:]

    @objc public func findPath(_ path: FPath) -> FNode? {
        if let value = value {
            return value.getChild(path)
        } else if let childKey = path.getFront(), !children.isEmpty {
            if let childTree = children[childKey] {
                return childTree.findPath(path.popFront())
            } else {
                return nil
            }
        } else {
            return nil
        }
    }

    @objc public func rememberData(_ data: FNode, onPath path: FPath) {
        if path.isEmpty {
            value = data
            children = [:]
        } else if let value = value {
            self.value = value.updateChild(path, withNewChild: data)
        } else {
            guard let childKey = path.getFront() else {
                return
            }
            let child = children[childKey, default: FSparseSnapshotTree()]
            child.rememberData(data, onPath: path.popFront())
            children[childKey] = child
        }
    }

    @objc public func forgetPath(_ path: FPath) -> Bool {
        if path.isEmpty {
            value = nil
            children = [:]
            return true
        } else if let value = value {
            if value.isLeafNode() {
                // non-empty path at leaf. the path leads to nowhere
                return false
            } else {
                let tmp = value
                self.value = nil
                tmp.enumerateChildren { key, node, stop in
                    self.rememberData(node, onPath: FPath(with: key))
                }
                // we've cleared out the value and set children. Call ourself
                // again to hit the next case
                return self.forgetPath(path)
            }
        } else if !children.isEmpty {
            let childKey = path.getFront()! // XXX TODO: FORCE UNWRAP IS SAFE HERE, BUT NOT NICE TO LOOK AT
            let p = path.popFront()
            if let child = children[childKey] {
                let safeToRemove = child.forgetPath(p)
                if safeToRemove {
                    children.removeValue(forKey: childKey)
                }
            }
            if children.isEmpty {
                return true
            } else {
                return false
            }
        } else {
            return true
        }
    }

    @objc public func forEachTreeAtPath(_ prefixPath: FPath, do closure: (FPath, FNode) -> Void) {
        if let value = value {
            closure(prefixPath, value)
        } else {
            forEachChild { key, tree in
                let path = prefixPath.child(fromString: key)
                tree.forEachTreeAtPath(path, do: closure)
            }
        }
    }

    @objc public func forEachChild(_ closure: (String, FSparseSnapshotTree) -> Void) {
        for (key, tree) in children {
            closure(key, tree)
        }
    }
}
